vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +207,206 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
+		return diag.Errorf("error waiting to finish creation of independent disk: %s", err)
 	}
 
 	diskHref := task.Task.Owner.HREF
 	disk, err := vdc.GetDiskByHref(diskHref)
 	if err != nil {
-		return fmt.Errorf("unable to find disk with href %s: %s", diskHref, err)
+		return diag.Errorf("unable to find disk with href %s: %s", diskHref, err)
 	}
 
 	d.SetId(disk.Disk.Id)
 
-	return resourceVcdIndependentDiskRead(d, meta)
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func resourceVcdIndependentDiskUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	if d.HasChanges("size_in_mb", "storage_profile", "description") {
+		_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+		if err != nil {
+			return diag.Errorf(errorRetrievingOrgAndVdc, err)
+		}
+
+		storageProfileValue := d.Get("storage_profile").(string)
+		var storageProfileRef *types.Reference
+
+		if storageProfileValue != "" {
+			storageReference, err := vdc.FindStorageProfileReference(storageProfileValue)
+			if err != nil {
+				return diag.Errorf("error finding storage profile %s", storageProfileValue)
+			}
+			storageProfileRef = &types.Reference{HREF: storageReference.HREF}
+		}
+
+		disk, err := vdc.GetDiskById(d.Id(), true)
+		if err != nil {
+			return diag.Errorf("error fetching independent disk: %s", err)
+		}
+
+		sliceOfVmsHrefs, err := disk.GetAttachedVmsHrefs()
+		if err != nil {
+			return diag.Errorf("error resourceVcdIndependentDiskUpdate faced issue fetching attached VMs")
+		}
+
+		// add global lock for shared disks to avoid deadlock possibility when different independent shared disks used by same VMs
+		lockGloballyIfNeeded(sliceOfVmsHrefs)
+		defer unlockGloballyIfNeeded(sliceOfVmsHrefs)
+		//lock VMs as another independent disk resource can be doing update with same VM
+		lockVms(sliceOfVmsHrefs)
+		defer unlockVms(sliceOfVmsHrefs)
+
+		diskDetailsForReAttach, diagErr := detachVms(vcdClient, disk, sliceOfVmsHrefs)
+		if diagErr != nil {
+			return diagErr
+		}
+
+		err = disk.Refresh()
+		if err != nil {
+			return diag.Errorf("error resourceVcdIndependentDiskUpdate error refreshing independent disk: %s", err)
+		}
+
+		disk.Disk.SizeMb = int64(d.Get("size_in_mb").(int))
+		disk.Disk.Description = d.Get("description").(string)
+		if storageProfileRef != nil {
+			disk.Disk.StorageProfile = storageProfileRef
+		}
+
+		task, err := disk.Update(disk.Disk)
+		if err != nil {
+			return diag.Errorf("error updating independent disk: %s", err)
+		}
+
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return diag.Errorf("error waiting to finish updating of independent disk: %s", err)
+		}
+
+		diagErr = attachBackVms(vcdClient, disk, diskDetailsForReAttach, sliceOfVmsHrefs)
+		if diagErr != nil {
+			return diagErr
+		}
+
+	}
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func lockGloballyIfNeeded(sliceOfVmsHrefs []string) {
```

Comment:
I am thinking how to improve code maintainability of this locking mechanism.
- Locking is one area where we need comments for both lock and unlock functions.
- And use more meaningful names.

```suggestion
// Add global lock for shared disks to avoid deadlock possibility when different independent shared disks used by same VMs		
func lockSharedDiskOpsGlobally(diskAttachedVmsHrefs []string) {
```

CC: @Didainius